src/reflect/scala/reflect/internal/names/ConcurrentMapNameTable.scala

```diff
@@ -2,11 +2,41 @@ package scala.reflect.internal.names
 
 import java.util.concurrent.ConcurrentHashMap
 
-class ConcurrentMapNameTable[T <: AnyRef](builder: (String) => T) extends NameTable[T] {
+class ConcurrentMapNameTable1[T <: AnyRef](builder: (String) => T) extends NameTable[T] {
 
   override def size: Int = data.size
 
   private val data = new ConcurrentHashMap[String, T]
 
   def find(source: String): T = data.computeIfAbsent(source, (s) => builder(s))
+
+  override def nonAllocatingStringLookup: Boolean = false
+}
+class ConcurrentMapNameTable2[T <: AnyRef](builder: (String) => T) extends NameTable[T] {
+
+  override def size: Int = data.size
+
+  private val data = new ConcurrentHashMap[String, T]
+  private val builderFn = new java.util.function.Function[String, T]{
+    override def apply(t: String): T = builder(t)
+  }
```

Comment:
@mkeskells In answer your question about this code: scala will statically hoist lambdas that don't capture anything. This one captures `this` (to call `this.builder`), so it isn't eligible for the automatic hoisting. You need to decide its worth the cost of an extra field in the instance to hoist -- it clearly makes sense here because relatively few `NameTable` instances are allocated/retained.